### PR TITLE
[Major] rename pass to test, move severity to be the last argument

### PR DESCRIPTION
**Pull request by ealush**
Breaking changes
---
* Moved severity to be the last argument to reduce api ambiguity

Non breaking changes
---
* Changed naming from `pass` to `test` to be more aligned with testing apis.

so basically, the change is from:
```js
passable('formName', specific, (pass) => {
    pass('fieldName', 'error string', 'warn', () => {})
});
```
to
```js
passable('formName', specific, (test) => {
    test('fieldName', 'error string', () => {}, 'warn')
});
```

Effects on consumers
------
The effect of both should not be large on existing consumers.

-- `warn` is not a common API, and since it was not mandatory, most consumers won't need to change anything to get the new API.
-- The name change has zero effect on consumption, as the function is not required, but given as an argument to the current passable instance.  Existing code will run exactly the same, but now the official recommendation is to use the argument as `test`.

### CODE REVIEW

File: documentation/enforce/README.md [hunk 1]
```diff
@@ -12,8 +12,8 @@ You can consume the `enforce` function in two ways:
 ```js
 import passable, {enforce} from 'passable';
 
-Passable('enforcement', (pass) => {
-    pass('test', 'enforce example', () => {
+Passable('enforcement', (test) => {
```
Comment (raphaelboukara) [inline documentation/enforce/README.md]: `Passable` -> `passable` right?

File: documentation/enforce/README.md [hunk 2]
```diff
@@ -22,16 +22,16 @@ Direct import is the newer standard, and is intended to support future functiona
 
 * `enforce` is the second argument of the passable function
 ```js
-Passable('enforcement', (pass, enforce) => {
-    pass('test', 'enforce example', () => {
+Passable('enforcement', (test, enforce) => {
```
Comment (raphaelboukara) [inline documentation/enforce/README.md]: same

File: documentation/getting_started/params.md [hunk 3]
```diff
@@ -4,7 +4,7 @@
 |------------|:---------:|:--------:|---------------------------------------------------------------------------|
 | `name`     | No        | String   | A name for the group of tests. E.G - form name                            |
 | `specific` | Yes       | Array    | Whitelist of tests to run.                 |
-| `passes`   | No        | Function | A function containing the actual validation logic.                        |
+| `tests`    | No        | Function | A function containing the actual validation logic.                        |
 | `custom`   | Yes       | Object   | Custom rules to extend the basic ruleset with. |
 
 ![api](../assets/img/passable-api.jpg)
```
Comment (raphaelboukara) [inline documentation/getting_started/params.md]: you need to update the picture

File: documentation/getting_started/writing_tests.md [hunk 4]
```diff
@@ -1,8 +1,8 @@
 # Writing tests
- Much like when writing unit-tests, writing validations with Passable is all about knowing in advance which values you expect to get. The structure is very similar to the familiar unit test `describe/it/expect` combo, only that with Passable the functions you will mostly run are `Passable/pass/enforce`.
+ Much like when writing unit-tests, writing validations with Passable is all about knowing in advance which values you expect to get. The structure is very similar to the familiar unit test `describe/it/expect` combo, only that with Passable the functions you will mostly run are `Passable/test/enforce`.
```
Comment (raphaelboukara) [inline documentation/getting_started/writing_tests.md]: why all the time you are wrinting `Passable` it's not `passable`?

File: documentation/test/how_to_fail.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# All ways to fail a `test`
+There are three ways of failing a test, and marking it as having a validation error:
+* Having an `enforce` function which ends up failing.
+```js
+test('field', 'should fail by enforce', () => {
+    enforce(1).allOf({ largerThan: 5 });
+});
```
Comment (raphaelboukara) [inline documentation/test/how_to_fail.md]: because enforce throwing an exception... so in fact there is 2 ways of failing a test, I don't think you have to speak about enforce here

Comment (raphaelboukara) [reply documentation/test/how_to_fail.md]: If you want you can add a note to give an example with chai or enforce but there is not 3 ways

File: documentation/test/specific.md [hunk 6]
```diff
@@ -1,21 +1,21 @@
...
+| `Object`        | Allows both setting `test` functions that should run, or not run
 
 **Remember** Using the array or string directly as your `specific` param is exactly the same as using `only`, and is provided as a shorthand to reduce clutter.
 
 ```js
-passable('formName', ['field_1', 'field_2'], (pass) => {...})
+passable('formName', ['field_1', 'field_2'], (test) => {...})
 // ------
-passable('formName', 'field_1', (pass) => {...})
+passable('formName', 'field_1', (test) => {...})
```
Comment (raphaelboukara) [inline documentation/test/specific.md]: and now passable with little p

File: documentation/test/specific.md [hunk 7]
```diff
@@ -1,21 +1,21 @@
-# Running or skipping `specific` passes
+# Running or skipping `specific` tests
 Sometimes you want to test only a specific field out of the whole form. For example, when validating upon user interaction (such as on input change), you probably do not want to validate all other fields as well. Similarly, you might want to **not** run the tests of specific fields, for example - fields that have not been touched by the user.
```
Comment (raphaelboukara) [inline documentation/test/specific.md]: I don't think it's a good idea to speak about `form`. Passable validate data, not important from where the data is coming

File: src/types.js [hunk 8]
```diff
@@ -6,8 +6,8 @@ declare type Rules = {
     [name: string]: Function
 };
 
-declare type Tests = {
-    [name: string]: AnyValue
+declare type CompoundTestObject = {
+    [rule: string]: AnyValue
```
Comment (raphaelboukara) [inline src/types.js]: why did you change that?

File: documentation/enforce/README.md (hunk 1 above)
Comment (ealush) [reply documentation/enforce/README.md]: This is a good comment, but not related to this specific PR. The docs issue needs to be addressed separately before pushing the next version.
👍 
Adding an issue.

File: src/types.js (hunk 8 above)
Comment (ealush) [reply src/types.js]: To reflect what's really inside.

**Comment by ealush** [comment]
@raphaelboukara, awesome comments about the documentation. Most of them are unrelated to the current PR, but should be taken care of nonetheless. I opened an issue containing your comments, and added it to the v6 roadmap. It will be fixed before pushing a new version.

File: src/Passable.js [hunk 9]
```diff
@@ -31,7 +31,7 @@ class Passable {
         return this.res;
     }
 
-    test = (fieldName: string, statement: string, ...args: [Severity, TestFn]) => {
+    test = (fieldName: string, statement: string, test: TestFn, severity: Severity) => {
```
Comment (raphaelboukara) [inline src/Passable.js]: where the `Severity` type is defined? why I can't find it in the top of the file, with all the imports?

File: documentation/test/warn_only_tests.md [hunk 10]
```diff
@@ -1,12 +1,12 @@
...
-By default, a failing `test` will set `hasValidationErrors` to `true`. Sometimes you would want to set a warn-only validation test (password strength, for example). In this case, you would add the `'warn'` flag to your test function.
+By default, a failing `test` will set `hasValidationErrors` to `true`. Sometimes you would want to set a warn-only validation test (password strength, for example). In this case, you would add the `'warn'` flag to your test function as the last argument.
 This will leave `hasValidationErrors` unchanged (other tests may have set it to `true`), and update `hasValidationWarnings` to `true`. It will also bump up `warnCount`.
 
 If no flag is added, your test function will default to `fail`.
 
 ```js
 Passable('WarnAndPass', (test) => {
-    test('WarnMe', 'Should warn and not fail', 'warn', () => false);
+    test('WarnMe', 'Should warn and not fail', () => false, 'warn');
```
Comment (raphaelboukara) [inline documentation/test/warn_only_tests.md]: (question not linked to this PR) Passable is a class, why you are not doing `new Passable` ? I understand that when the consumer write `Passable('dataName', (test) => {...})` the `test` function is define [here](https://github.com/fiverr/passable/blob/90b45126270af0b0e69dbe16a8c18102508b8ae2/src/Passable.js#L34) right? But I can't find where this: `Passable('dataName', (test) => {...})` magic happen?

File: src/Passable.js [hunk 11]
```diff
@@ -40,21 +40,15 @@ class Passable {
...
 
-        this.res.initFieldCounters(fieldName);
-
-        const isValid: boolean = testRunner(callback);
+        const isValid: boolean = testRunner(test);
 
         if (!isValid) {
-            const severity: Severity = lastIndex !== 0 ? args[0] : FAIL;
             this.res.fail(fieldName, statement, severity);
         }
```
Comment (raphaelboukara) [inline src/Passable.js]: in fact you can do `testRunner(test) && this.res.fail(fieldName, statement, severity);`

File: src/Passable.js [hunk 12]
```diff
@@ -40,21 +40,15 @@ class Passable {
             return;
         }
 
-        const lastIndex: number = args.length - 1;
-        let callback: Function;
+        this.res.initFieldCounters(fieldName);
```
Comment (raphaelboukara) [inline src/Passable.js]: Why `initFieldCounters` happens for each `test`, maybe you want to say `pushToFieldCounters` ? What I missed?

File: src/Passable.js [hunk 13]
```diff
@@ -40,21 +40,15 @@ class Passable {
...
 
-        if (typeof args[lastIndex] === 'function') {
-            callback = args[lastIndex];
-        } else {
+        if (typeof test !== 'function') {
```
Comment (raphaelboukara) [inline src/Passable.js]: Why this not the first check int this function? Why you want to `initFieldCounters` if there is nothing to test?

File: src/Passable.js [hunk 14]
```diff
@@ -31,7 +31,7 @@ class Passable {
         return this.res;
```
Comment (raphaelboukara) [inline src/Passable.js]: at the top of this file you have `const FAIL: Severity = 'fail';` you are not exporting this const and it seems nobody use it in the file, maybe you can remove it?

File: src/result_object/helpers/fail/index.js [hunk 15]
```diff
@@ -1,7 +1,8 @@
 // @flow
 import PassableResponse from '../../index';
 
-export const WARN: string = 'warn';
+export const WARN: Severity = 'warn';
+export const FAIL: Severity = 'fail';
```
Comment (raphaelboukara) [inline src/result_object/helpers/fail/index.js]: But why you don't have to import `Severity` 😣, is `Severity` a flow type?

File: src/result_object/helpers/fail/index.js [hunk 16]
```diff
@@ -1,7 +1,8 @@
...
 
 type CountName = 'failCount' | 'warnCount';
```
Comment (raphaelboukara) [inline src/result_object/helpers/fail/index.js]: line 12: `const isFail: boolean = (severity !== WARN);` why not `const isFail: boolean = (severity === FAIL);` ?

File: src/result_object/helpers/fail/index.js [hunk 17]
```diff
@@ -1,7 +1,8 @@
...
 type ObjectName = 'validationErrors' | 'validationWarnings';
```
Comment (raphaelboukara) [inline src/result_object/helpers/fail/index.js]: I don't understand why you enrich the `this` and return it instead of return a new object, can we sit together?

Comment (raphaelboukara) [reply src/result_object/helpers/fail/index.js]: is `this` the res of Passable??

File: src/result_object/index.js [hunk 18]
```diff
@@ -64,7 +64,7 @@ class ResultObject {
      * @param {string} severity - Whether it is a `fail` or `warn` test.
      * @return {Function} Calling a helper function
      */
-    fail(fieldName: string, statement: string, severity: Severity) {
+    fail(fieldName: string, statement: string, severity: Severity = FAIL) {
         return fail.call(this, fieldName, statement, severity);
```
Comment (raphaelboukara) [inline src/result_object/index.js]: The famous `this`

Comment (raphaelboukara) [reply src/result_object/index.js]: Why to define the default severity here and not [here](https://github.com/fiverr/passable/blob/90b45126270af0b0e69dbe16a8c18102508b8ae2/src/Passable.js#L34) ?

Comment (ealush) [reply src/result_object/index.js]: solved by itself :)

File: src/result_object/helpers/fail/index.js (hunk 17 above)
Comment (ealush) [reply src/result_object/helpers/fail/index.js]: Good point. Moving fail into the result object class.

File: src/result_object/helpers/fail/index.js (hunk 16 above)
Comment (ealush) [reply src/result_object/helpers/fail/index.js]: Adding WARN and FAIL exports, and checking explicit severity.
`const severity = [WARN, FAIL].includes(userSeverity) ? userSeverity : FAIL;`

File: documentation/test/warn_only_tests.md [hunk 19]
```diff
@@ -1,12 +1,30 @@
...
+const passable = require('passable');
+const WARN = passable.WARN;
+```
+
+Use it like this:
+
+```js
+import passable, { WARN, enforce } from 'passable';
+
+passable('WarnAndPass', (test) => {
```
Comment (raphaelboukara) [inline documentation/test/warn_only_tests.md]: you missed `specific`

File: src/result_object/index.js [hunk 20]
```diff
@@ -1,9 +1,8 @@
...
 
-declare type ErrorAndWarningObject = {
-    [name: string]: Array<string>
-};
+export const WARN: Severity = 'warn';
+export const FAIL: Severity = 'fail';
+const severities: Array<Severity> = [WARN, FAIL];
 
 /** Class representing validation state. */
 class ResultObject {
```
Comment (raphaelboukara) [inline src/result_object/index.js]: don't need to return this in constructor

File: src/result_object/index.js [hunk 21]
```diff
@@ -62,10 +61,29 @@ class ResultObject {
...
+    fail(fieldName: string, statement: string, severity: Severity) {
+        const selectedSeverity: Severity = severity && severities.includes(severity) ? severity : FAIL;
+        const isWarn: boolean = selectedSeverity === WARN;
+
+        const countName: string = isWarn ? 'warnCount' : 'failCount';
+
+        if (isWarn) {
+            this.hasValidationWarnings = true;
+            this.validationWarnings[fieldName] = this.validationWarnings[fieldName] || [];
+            this.validationWarnings[fieldName].push(statement);
```
Comment (raphaelboukara) [inline src/result_object/index.js]: I think it will be clearer to juxtapose 
```js
this.testsPerformed[fieldName].warnCount++;
this.warnCount++;
```

or maybe you can create a function for each of them: `bumpTestCounter`, `bumpWarnCounter`, `bumpFailCounter`

File: src/result_object/index.js [hunk 22]
```diff
@@ -62,10 +61,29 @@ class ResultObject {
...
+        } else {
+            this.hasValidationErrors = true;
+            this.validationErrors[fieldName] = this.validationErrors[fieldName] || [];
+            this.validationErrors[fieldName].push(statement);
+            this.failCount++;
+        }
+
+        this.testsPerformed[fieldName][countName]++;
+
+        return this;
```
Comment (raphaelboukara) [inline src/result_object/index.js]: I don't like the functions `getErrors`, do you think it's possible to add a getter to `this.errors` and another getter (more complicated) on `this[fieldName].errors`. Same for warnings, I think it's very nice for the user to simply do `res[fieldName].errors` or `res[fieldName].warnings`. I don't think you need the nested `validationErrors` and `validationWarnings`.

Comment (raphaelboukara) [reply src/result_object/index.js]: The question is how to do that?
Maybe you can use `Proxy`? something like
```js
new Proxy(passable.res, {
    get: function(target, name, receiver) {
        switch(name) {
            case 'errors':
                return this.getErrors(target);
            case 'warnings':
                return this.getWarnings(target);
            default:
                return target[name];
        }
     }
});
```
outside the class, when creating the passable instance.

Comment (raphaelboukara) [reply src/result_object/index.js]: no no, not a good idea it won't work at all

Comment (raphaelboukara) [reply src/result_object/index.js]: azov

File: src/result_object/index.js (hunk 20 above)
Comment (ealush) [reply src/result_object/index.js]: very true

File: src/result_object/index.js (hunk 22 above)
Comment (ealush) [reply src/result_object/index.js]: >I don't like the functions getErrors, do you think it's possible to add a getter to this.errors and another getter (more complicated) on this[fieldName].errors. Same for warnings, I think it's very nice for the user to simply do res[fieldName].errors or res[fieldName].warnings. I don't think you need the nested validationErrors and validationWarnings.

I actually think it simplifies working with passable.
Think about cases that I have multiple fields with errors.

```js
const result = passable(...);
```

now, to display the errors, I would need to do something like this:

```js
const usernameErr = res.validationErrors.username.length && res.validationErrors.username[0];
const passwordErr = res.validationErrors.password.length && res.validationErrors.password[0];
const emailErr = res.validationErrors.email.length && res.validationErrors.email[0];
```

This is basically a shortcut that does the same thing.

File: src/result_object/index.js [hunk 23]
```diff
@@ -56,6 +55,32 @@ class ResultObject {
...
+    }
+
+    /**
+     * Bumps field's error counts and adds error string
+     * @param {string} fieldName - The name of the field.
+     * @param {string} statement - The error string to add to the object.
+     */
+    bumpTestError(fieldName: string, statement: string) {
+        this.hasValidationErrors = true;
+        this.validationErrors[fieldName] = this.validationErrors[fieldName] || [];
```
Comment (raphaelboukara) [inline src/result_object/index.js]: same here `this.validationErrors[fieldName] ||= [];`

File: src/result_object/index.js [hunk 24]
```diff
@@ -56,6 +55,32 @@ class ResultObject {
...
+        this.validationErrors[fieldName].push(statement);
+        this.failCount++;
+        this.testsPerformed[fieldName].failCount++;
+    }
```
Comment (raphaelboukara) [inline src/result_object/index.js]: I am not sur about all those assignment: 
. `failCount` can be the result of a function
. `testsPerformed[fieldName].failCount` same
. `hasValidationErrors` same

This will remove from code a lot of assigment, what do you think? If you are afraid about perf, you can maybe add a benchmark

File: src/result_object/index.js [hunk 25]
```diff
@@ -64,24 +89,13 @@ class ResultObject {
...
-            this.hasValidationErrors = true;
-            this.validationErrors[fieldName] = this.validationErrors[fieldName] || [];
-            this.validationErrors[fieldName].push(statement);
-            this.failCount++;
-        }
 
-        this.testsPerformed[fieldName][countName]++;
+        selectedSeverity === WARN
+            ? this.bumpTestWarning(fieldName, statement)
+            : this.bumpTestError(fieldName, statement);
```
Comment (raphaelboukara) [inline src/result_object/index.js]: cool

File: src/result_object/index.js (hunk 24 above)
Comment (raphaelboukara) [reply src/result_object/index.js]: If all this properties are equal to `f(validationErrors[fieldName])` that's mean you just need to unit test those functions and not have to test the entire object

File: src/result_object/index.js [hunk 26]
```diff
@@ -56,6 +55,32 @@ class ResultObject {
...
     }
 
+    /**
+     * Bumps field's warning counts and adds warning string
+     * @param {string} fieldName - The name of the field.
+     * @param {string} statement - The error string to add to the object.
+     */
+    bumpTestWarning(fieldName: string, statement: string) {
+        this.hasValidationWarnings = true;
+        this.validationWarnings[fieldName] = this.validationWarnings[fieldName] || [];
```
[the remark this answers is not in the document]
Comment (ealush) [reply src/result_object/index.js]: That's not really a thing. Wish it was though.

File: src/result_object/index.js (hunk 24 above)
Comment (ealush) [reply src/result_object/index.js]: 1. Regarding the dynamic enumeration of the counters. Really cool idea, outside the scope of this change. I don't worry about the performance. The only issue I can think of is that there will be much more logic to maintain and test. This one is as simple as it gets: `++`.

2. I prefer testing the whole Object and not just single property. I want to know that everything works together and I don't get any weird side effects.
